File: v5_dependency_inversion/python/car_rental_agency.py
class BMW:
    def rent_bmw(self, model: str):
        print(f'BMW rented {model}')

class Mercedes:
    def rent_mercedes(self, model: str):
        print(f'Mercedes rented {model}')

class CarRentalAgency:
    def rent_car(self, brand: str, model: str):
        if brand == 'BMW':
            BMW().rent_bmw(model)
        elif brand == 'Mercedes':
            Mercedes().rent_mercedes(model)

agency = CarRentalAgency()
agency.rent_car('BMW', 'X5')
agency.rent_car('Mercedes', 'GLE')

File: v5_dependency_inversion/java/CarRentalAgency.java
public class CarRentalAgency {

    public static void main(String[] args) {
        CarRentalAgency agency =  new CarRentalAgency();
        agency.rentCar("BMW", "X5");
        agency.rentCar("Mercedes", "GLE");
    }

    public void rentCar(String brand, String model) {
        if (brand == "BMW") {
            new BMW().rentBMW(model);
        } else if (brand == "Mercedes") {
            new Mercedes().rentMercedes(model);
        }
    }
}

class BMW {

    public void rentBMW(String model) {
        System.out.println("BMW rented " + model);
    }
}


class Mercedes {

    public void rentMercedes(String model) {
        System.out.println("Mercedes rented " + model);
    }
}
